### PR TITLE
Fix backwards compatibility issue in AWS provider's _get_credentials

### LINKED ISSUE #20457: Release 2.5.1 of Amazon Provider
### Apache Airflow version

2.2.3 (latest released)

### What happened

The 2.5.0 version of Amazon Provider contains breaking changes - https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/commits.html specifically https://github.com/apache/airflow/commit/83b51e53062dc596a630edd4bd01407a556f1aa6 or the combination of the following:

- https://github.com/apache/airflow/commit/83b51e53062dc596a630edd4bd01407a556f1aa6
- https://github.com/apache/airflow/commit/d58df468c8d77c5d45e80f2333eb074bb7771a95
- https://github.com/apache/airflow/commit/4be04143a5f7e246127e942bf1d73abcd22ce189

and confirmed by @uranusjr 

I have yanked 2.5.0 and we will need to release ~3.0.0~ 2.5.1 with backwards compatibility fixes 

I have updated the constraints for 2.2.3 for now - https://github.com/apache/airflow/commit/62d490d4da17e35d4ddcd4ee38902a8a4e9bbfff

UPDATED: (@potiuk) to reflect that we are going to release 2.5.1 instead of 3.0.0

### PR DESCRIPTION
**Pull request by potiuk**
The #19815 change introduced backwards incompatibility for
the _get_credentials method - which is a centerpiece of AWS
provider and is likely to be overwritten by the user who want
for example inject auditing or other credentials-related custom
beheviours when interfacing with AWS even if the method is
protected.

The change added default for region, which caused signature
incompatibility with such derived classes. Unfortunately, we
already released 2.5.0 provider with this change. We had to
yank it and in order to avoid adding backwards-incompatible
3.0.0 release we are going to release 2.5.1 with this change
included.

Fixes: #20457

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
Once we review and merge this one I will prepare 2.5.1 release of the provider.

**Comment by potiuk** [comment]
All `test_base_aws.py` passed locally after the changes.

**Review by kaxil — APPROVED** [review]
lgtm but would wait till @dstandish or @uranusjr reviews it

**Comment by potiuk** [comment]
Aaaaah tooo fast :) (just saw your comment @kaxil  @dstandish @uranusjr  - let me know if there is anything to fix :)

**Comment by dstandish** [comment]
nothing to fix i'd say but i am curious why we no longer let the hook handle auth more completely.... 

like it used to be 

```python
        s3_resource = self.get_resource_type('s3')
        return s3_resource.Bucket(bucket_name)
```

but as of [this change](https://github.com/potiuk/airflow/commit/83b51e53062dc596a630edd4bd01407a556f1aa6) we have much more complicated expression and get our hands more dirty.  just curious  for the reasoning on that.  is there a reason not to let base hook handle giving us an authenticated resource, instead of repeating that logic within method calls?

also why _not_ let region_name be optional (i.e. have a default of None) in `_get_credentials`?  that doesn't necessarily harm backcompat

**Comment by potiuk** [comment]
You'll find answers here @dstandish : https://github.com/apache/airflow/issues/20457#issuecomment-999185822

The backcompat is kinda unexpected.  But somewhat realistic case (and I guess @kaxil and @uranusjr already got a report about it) when someone overrides the _get_credentials() "an old way" and expand_role() would fail in this case. 

This is not obvious but might happen.

**Comment by potiuk** [comment]
But yeah. In this case when I added the default to expand_role - it could be back set to None default.

**Pull request merged**
